include/envoy/http/codec.h

```diff
@@ -436,7 +437,7 @@ struct Http1Settings {
   };
 
   // How header keys should be formatted when serializing HTTP/1.1 headers.
-  HeaderKeyFormat header_key_format_{HeaderKeyFormat::Default};
+  absl::variant<HeaderKeyFormat, StatefulHeaderKeyFormatterFactoryPtr> header_key_format_;
```

Comment:
WDYT about having PreserveCase in the HeaderKeyFormat enum, and then having an instance of StatefulHeaderKeyFormatterFactoryPtr as a separate member variable, without the absl::variant, if we need that factory.

In terms of memory I think that will be less, and in terms of readability of call-sites I suspect it might be more intuitive to have PreserveCase be a sibling option to Default and ProperCase.

I think the advantage of the factory pattern you have here it is that we could have some stateful formatter that does something other than PreserveCase. Is that something we anticipating having a need for?

Reply:
> I think the advantage of the factory pattern you have here it is that we could have some stateful formatter that does something other than PreserveCase. Is that something we anticipating having a need for?

I went ahead and did the work to make this an extension point which is why I did it this way. Arguably, I could deprecate proper case and rewrite it as a "stateful" extension which would simplify all of this. IMO it's better to make this an extension because I would rather not be back here again in the future adding yet another type of formatter, but I don't feel strongly about. I think the options are:
1) Leave it like it is
2) Remove the extension point and just hard code preserve case with good documentation.
3) Deprecate proper case and make it an extension.

WDYT?

Reply:
+1 to Snow's plan. I think that would be cleaner and just as general.

Reply:
> I think we can retain the extension mechanism but improve the code readability by making StatefulHeaderFormatter a part of the HeaderKeyFormat enum without changing the proto enum. We can then make use of the StatefulHeaderKeyFormatterFactoryPtr when header_key_format_ is StatefulHeaderFormatter without all the variant acrobatics.

IMO the variant is better, because it actually handles the either or very well, but if you all feel strongly about it I can change it.

Reply:
> +1 to Snow's plan. I think that would be cleaner and just as general.

OK sounds good. Like I said I like the variant as it forces the construct, but I'm not going to die on this hill. :)